### PR TITLE
[RLlib] Issue 14119: Fix TD3 policy delay for torch.

### LINKED ISSUE #14119: [RLLIB] pytorch TD3 "policy_delay" option ignored.
### What is the problem?
The policy_delay configuration option of TD3 is ignored in the pytorch version. The reason is that the DDPGTrainer is customized from GenericOffPolicyTrainer which uses TrainOneStep in its execution schedule. TrainOneStep calls torch_policy.py:learn_on_batch(...) which steps the opsimizers directly.

https://github.com/ray-project/ray/blob/4846a6c2d04c4a71cf9619acab46ec82dec3f2b6/rllib/policy/torch_policy.py#L354-L360

DDPG attempts to implement the policy_delay by defining a custom apply_gradients method but since learn_on_batch steps the optimizers directly apply_gradients is not called. 

https://github.com/ray-project/ray/blob/4846a6c2d04c4a71cf9619acab46ec82dec3f2b6/rllib/agents/ddpg/ddpg_torch_policy.py#L192-L201

https://github.com/ray-project/ray/blob/4846a6c2d04c4a71cf9619acab46ec82dec3f2b6/rllib/agents/ddpg/ddpg_torch_policy.py#L270-L290

*Ray version and other system information (Python version, TensorFlow version, OS):*
python: 3.x
rlllib: master

### Reproduction (REQUIRED)
Please provide a short code snippet (less than 50 lines if possible) that can be copy-pasted to reproduce the issue. The snippet should have **no external library dependencies** (i.e., use fake or mock data / environments):

Here is a script that will show it being called correctly in tf1 but is not called at all in pytorch. TF2 works too but I commented it out because it it was slow.
```
import ray
from ray.rllib.agents.ddpg import DDPGTrainer
from ray.rllib.agents.ddpg.ddpg_tf_policy import DDPGTFPolicy
from ray.rllib.agents.ddpg.ddpg_torch_policy import DDPGTorchPolicy, apply_gradients_fn as ddpg_apply_grad_fn
from ray.rllib.agents.ddpg.td3 import TD3_DEFAULT_CONFIG, TD3Trainer
from ray.rllib.utils.framework import try_import_tf,get_variable
tf1, tf, tfv = try_import_tf()


def build_apply_op_tf(policy, optimizer, grads_and_vars):
    # For policy gradient, update policy net one time v.s.
    # update critic net `policy_delay` time(s).
    should_apply_actor_opt = tf.equal(
        tf.math.floormod(policy.global_step, policy.config["policy_delay"]), 0)

    def make_apply_op():
        return policy._actor_optimizer.apply_gradients(
            policy._actor_grads_and_vars)

    actor_op = tf.cond(
        should_apply_actor_opt,
        true_fn=make_apply_op,
        false_fn=lambda: tf.no_op())
    critic_op = policy._critic_optimizer.apply_gradients(
        policy._critic_grads_and_vars)
    # Increment global step & apply ops.
    if policy.config["framework"] in ["tf2", "tfe"]:
        print("Using build_apply_op in tf2")
        policy.global_step.assign_add(1)
        return tf.no_op()
    else:
        with tf1.control_dependencies([tf1.assign_add(policy.global_step, 1),tf.print("Using build_apply_op in tf1")]):
            return tf.group(actor_op, critic_op)

def apply_gradients_fn(policy, gradients):
    print("Called apply_gradients_fn")
    ddpg_apply_grad_fn(policy, gradients)

TD3_TF_APPLY_TEST=DDPGTFPolicy.with_updates(apply_gradients_fn=build_apply_op_tf)
TD3_TORCH_APPLY_TEST=DDPGTorchPolicy.with_updates(apply_gradients_fn=apply_gradients_fn)

def get_policy_class(config):
    if config["framework"] == "torch":
        return TD3_TORCH_APPLY_TEST
    else:
        return TD3_TF_APPLY_TEST

DDPG_APPLY_TRAINER = TD3Trainer.with_updates(
        default_policy=TD3_TF_APPLY_TEST,
        get_policy_class=get_policy_class,
    )

def main():
    ray.init(local_mode=True, num_cpus=1)

    td3_config = TD3_DEFAULT_CONFIG.copy()
    # === Model ===
    td3_config["actor_hiddens"]=[64, 64]
    td3_config["critic_hiddens"]=[64, 64]

    print("#"*10, " Testing TF1 ", "#"*10)
    td3_config["framework"]="tf"
    trainer = DDPG_APPLY_TRAINER(config=td3_config, env="Pendulum-v0")
    for _ in range(2):
         trainer.train()
    trainer.stop()

    #print()
    #print("#"*10, " Testing TF2 ", "#"*10)
    #td3_config["framework"]="tf2"
    #trainer = DDPG_APPLY_TRAINER(config=td3_config, env="Pendulum-v0")
    #for _ in range(2):
    #    trainer.train()
    #trainer.stop()

    print()
    print("#"*10, " Testing TORCH ", "#"*10)
    td3_config["framework"]="torch"
    trainer = DDPG_APPLY_TRAINER(config=td3_config, env="Pendulum-v0")
    for _ in range(2):
        trainer.train()
    print("Finished train loop in pytorch.")
    trainer.stop()
    ray.shutdown()

if __name__ == '__main__':
    main()
```

If the code snippet cannot be run by itself, the issue will be closed with "needs-repro-script".

- [x] I have verified my script runs in a clean environment and reproduces the issue.
- [x] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

### PR DESCRIPTION
**Pull request by mvindiola1**
Fix DDPG edge cases due to multiple optimizers.

## Why are these changes needed?
These commits fix two issues with DDPG
* Torch version not using policy_delay for TD3
The torch policy was not calling the special cased apply_gradients function because of learn_on_batch. This commit adds a special fused case that calls apply gradients 
* eager_tf_policy get_state() does not check to make sure that policy._optimizer is not None. This fails when saving state for DDPG because it has has special variables for the optimizers policy._actor_optimizer and policy._critic_optimizer

## Related issue number
Closes #14119

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by sven1977 — APPROVED** [review]
Thanks for this fix @mvindiola1 !
